### PR TITLE
fix: display earned course credentials even if the grade is missing

### PR DESCRIPTION
**Pull request by justinhynes**
There are well documented communication issues between the LMS and Credentials systems. There are times that Credentials will just miss a grade update from the LMS. The grades stored in Credentials are used to render a learner's progress in their courses through the Learner Record UI.

This PR changes the behavior of the API that feeds data to the Learner Record MFE (program record page) and the Support Tools.

Previously, when rendering the program record page, if there are no grades available for a course (that is part of a program) we would show the course credential as "unearned".

This has caused a lot of confusion for our learners, as a missing grade in Credentials does not mean they haven't earned the certificate. In fact, most of the time, the learner has been able to view and share their course certificate and they are confused as to why it shows as "unearned" on their program record(s).

Now, if the system is aware of an earned Credential, we will populate the issue_date field (which the frontend uses to determine if a certificate is earned or not). This way we can reflect the learner record as accurately as possible for a learner.

**Screenshots**
_Current, a learner who has earned a credential in a course, but the grade is missing, will see the credential as "Not Earned"_
image

_After change, a learner who has earned a credential in a course, but the grade is missing, will see the credential as earned (albeit missing grade information)_
image

**Run JavaScript tests locally with Karma**

There is work being done on a fix to get Karma to run in CI. Until then, however, contributors are required to run these tests locally.

- [x] Make sure you are inside the devstack container
- [x] Run `make test-karma`
- [x] All tests pass

### CODE REVIEW

File: credentials/apps/records/tests/test_api.py [hunk 1]
```diff
@@ -184,26 +207,236 @@ def test_get_transformed_pathway_data(self):
...
+        self._assert_results(expected_highest_attempt_dict, highest_attempt_dict)
+        assert (
+            float(highest_attempt_dict.get(self.course1).percent_grade) == self.course1_courserun2_grade.percent_grade
+        )
+        assert (
+            float(highest_attempt_dict.get(self.course2).percent_grade) == self.course2_courserun1_grade.percent_grade
+        )
+        assert expected_issue_date_course2 == last_updated
+
+    def test_get_transformed_grade_data_has_credential_no_grade(self):
```
Comment (justinhynes) [inline credentials/apps/records/tests/test_api.py:252]: This test (and the test following this one) are not directly related to the changes in this PR. 

They _are_ important to the functionality of the feature/component as a whole. Since I was already adding/changing existing tests, I spent some time to add additional tests for edge case functionality that want to keep tabs on.

File: credentials/apps/records/tests/test_api.py [hunk 2]
```diff
@@ -184,26 +207,236 @@ def test_get_transformed_pathway_data(self):
...
+        self._assert_results(expected_result_course2, result[1])
         self._assert_results(expected_highest_attempt_dict, highest_attempt_dict)
-        assert float(highest_attempt_dict.get(self.course).percent_grade) == self.grade_high.percent_grade
-        assert expected_issue_date == last_updated
+        assert (
+            float(highest_attempt_dict.get(self.course1).percent_grade) == self.course1_courserun2_grade.percent_grade
+        )
+        assert expected_issue_date_course1 == last_updated
+
+    def test_get_transformed_grade_data_earned_credential_with_visible_date(self):
```
Comment (justinhynes) [inline credentials/apps/records/tests/test_api.py:341]: This test (and the test following this one) are not directly related to the changes in this PR. 

They _are_ important to the functionality of the feature/component as a whole. Since I was already adding/changing existing tests, I spent some time to add additional tests for edge case functionality that we want to keep tabs on.

File: credentials/apps/records/api.py [hunk 3]
```diff
@@ -185,37 +185,73 @@ def _get_transformed_grade_data(program, user):
...
+                    "issue_date": issue_date.isoformat(),
+                    "percent_grade": float(grade.percent_grade),
+                    "letter_grade": grade.letter_grade or _("N/A"),
                 }
             )
             added_courses.add(course)
-        elif grade is not None and grade.course_run == course_run:
-            user_credential = user_credential_dict.get(course_run.key)
-            issue_date = get_credential_dates(user_credential, False)
+        # Case 2: Learner has earned a credential in, but we have no record of a grade for, a course-run of a course
```
Comment (deborahgu) [inline credentials/apps/records/api.py:221]: does it actually make sense to build these three dicts separately? They seem similar enough that it seems like it would make more sense to create a variable for `attempts`, `letter_grade` etc. for the cases and just build the  dict once.

Comment (justinhynes) [reply credentials/apps/records/api.py:221]: OK! I can try and make it a bit more DRY :)

File: credentials/apps/records/api.py [hunk 4]
```diff
@@ -117,7 +117,7 @@ def _get_transformed_pathway_data(program, user):
     return pathway_data
 
 
-def _get_transformed_grade_data(program, user):
+def _get_transformed_grade_data(program, user):  # pylint: disable=too-many-statements
```
Comment (deborahgu) [inline credentials/apps/records/api.py:120]: oof. I'm reading this and it looks clearer to me than the other way, but arguably anything that triggers movement into `too-many-statements` is a red flag, you know? So the DRY improvements are balanced out by the `too-many-statements` trigger. I'm going to go ahead and approve this, but it sounded like you thought the other one was more clear to read, so given the above if you want to roll back to the old way I will go ahead and approve that instead.

**Review by deborahgu — APPROVED** [review]

File: credentials/apps/records/api.py (hunk 4 above)
Comment (justinhynes) [reply credentials/apps/records/api.py:120]: The linting warning came after the most recent changes. 

I think this code is terribly overcomplicated either way, and it really requires a more extensive re-write, so I'm trying to pick the least horrible improvement (for now)
